File: Math/BinaryExponentiation/374/374.cpp
#include <iostream>
using namespace std;

long long p(int B, int P, int M)
{
    if (P == 0)
        return 1 % M;
    if (P == 1)
        return B;
    long long res = p(B, P >> 1, M) % M;
    res = (res % M) * (res % M);
    res %= M;
    if (P & 1)
    {
        res *= B % M;
        res %= M;
    }
    return res;
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    int B, P, M;
    while(cin >> B >> P >> M)
    {
        cout << p(B % M, P, M) << '\n';
    }
}

File: Math/BinaryExponentiation/374/374.py
def p(B, P, M):
    if (P == 0):
        return 1 % M
    if (P == 1):
        return B
    res = p(B, P // 2, M) % M
    res = (res % M) * (res % M)
    res %= M
    if (P % 2 == 1):
        res *= B % M
        res %= M
    return res

while (1):
    try:
        B = int(input())
        P = int(input())
        M = int(input())
        print(p(B % M, P, M))
        input()
    except EOFError:
        break